File: truco_bot/core/state.py
from dataclasses import dataclass, field

from truco_bot.core.actions import Action
from truco_bot.core.card import Card
from truco_bot.core.rules import (
    calculate_envido,
    calculate_falta_envido_points,
    compare_cards,
    resolve_hand,
)


@dataclass(slots=True)
class GameState:
    hands: list[list[Card]]
    mano: int
    active_player: int
    score_p0: int = 0
    score_p1: int = 0
    max_score: int = 30
    current_trick: int = 0
    trick_cards: list[tuple[int, Card]] = field(default_factory=list)
    trick_results: list[int] = field(default_factory=list)
    played_cards: list[tuple[int, Card]] = field(default_factory=list)
    trick_leader: int = 0

    envido_resolved: bool = False
    envido_chain: list[Action] = field(default_factory=list)
    pending_envido_response: bool = False
    pending_envido_from: int | None = None

    truco_level: int = 1
    truco_caller: int | None = None
    pending_truco_response: bool = False
    pending_truco_from: int | None = None

    is_hand_done: bool = False
    winner: int | None = None
    points_won: int = 0


def create_initial_hand_state(
    hands: list[list[Card]],
    mano: int = 0,
    score_p0: int = 0,
    score_p1: int = 0,
    max_score: int = 30,
) -> GameState:
    return GameState(
        hands=hands,
        mano=mano,
        active_player=mano,
        score_p0=score_p0,
        score_p1=score_p1,
        max_score=max_score,
        trick_leader=mano,
    )


_EMPTY_ACTIONS: list[Action] = []
_ENVIDO_RESPONSES: list[Action] = [Action.QUIERO_ENVIDO, Action.NO_QUIERO_ENVIDO]
_ENVIDO_CALL_ENVIDO: list[Action] = [
    Action.QUIERO_ENVIDO,
    Action.NO_QUIERO_ENVIDO,
    Action.ENVIDO,
    Action.REAL_ENVIDO,
    Action.FALTA_ENVIDO,
]
_ENVIDO_CALL_REAL: list[Action] = [
    Action.QUIERO_ENVIDO,
    Action.NO_QUIERO_ENVIDO,
    Action.FALTA_ENVIDO,
]
_TRUCO_RESPONSES_1: list[Action] = [
    Action.QUIERO_TRUCO,
    Action.NO_QUIERO_TRUCO,
    Action.IR_AL_MAZO,
]
_TRUCO_RESPONSES_2: list[Action] = [
    Action.QUIERO_TRUCO,
    Action.NO_QUIERO_TRUCO,
    Action.IR_AL_MAZO,
    Action.VALE_CUATRO,
]
_TRUCO_RESPONSES_3: list[Action] = [
    Action.QUIERO_TRUCO,
    Action.NO_QUIERO_TRUCO,
    Action.IR_AL_MAZO,
]


def get_legal_actions(state: GameState) -> list[Action]:
    if state.is_hand_done:
        return _EMPTY_ACTIONS.copy()

    player = state.active_player

    if state.pending_envido_response:
        last_call = state.envido_chain[-1] if state.envido_chain else None
        if last_call == Action.ENVIDO:
            if state.envido_chain.count(Action.ENVIDO) >= 2:
                return [Action.QUIERO_ENVIDO, Action.NO_QUIERO_ENVIDO, Action.REAL_ENVIDO, Action.FALTA_ENVIDO]
            return _ENVIDO_CALL_ENVIDO.copy()
        elif last_call == Action.REAL_ENVIDO:
            return _ENVIDO_CALL_REAL.copy()
        return _ENVIDO_RESPONSES.copy()

    if state.pending_truco_response:
        if state.truco_level == 2:
            return _TRUCO_RESPONSES_2.copy()
        elif state.truco_level == 3:
            return _TRUCO_RESPONSES_3.copy()
        return _TRUCO_RESPONSES_1.copy()

    actions = []
    # envido
    if (
        state.current_trick == 0
        and not state.envido_resolved
        and len(state.trick_cards) < 2
        and len(state.envido_chain) == 0
    ):
        actions.extend([Action.ENVIDO, Action.REAL_ENVIDO, Action.FALTA_ENVIDO])

    # truco
    if state.truco_level == 1:
        actions.append(Action.TRUCO)
    elif state.truco_level == 2 and state.truco_caller != player:
        actions.append(Action.RETRUCO)
    elif state.truco_level == 3 and state.truco_caller != player:
        actions.append(Action.VALE_CUATRO)

    # card plays
    hand = state.hands[player]
    if len(hand) > 0:
        actions.append(Action.PLAY_CARD_0)
        actions.append(Action.PLAY_CARD_DOWN_0)
    if len(hand) > 1:
        actions.append(Action.PLAY_CARD_1)
        actions.append(Action.PLAY_CARD_DOWN_1)
    if len(hand) > 2:
        actions.append(Action.PLAY_CARD_2)
        actions.append(Action.PLAY_CARD_DOWN_2)

    actions.append(Action.IR_AL_MAZO)
    return actions


# ponytail: fast shallow clone replaces copy.deepcopy for 30x throughput
def _clone_state(state: GameState) -> GameState:
    hands_copy = (
        [state.hands[0].copy(), state.hands[1].copy()]
        if len(state.hands) == 2
        else [h.copy() for h in state.hands]
    )
    return GameState(
        hands_copy,
        state.mano,
        state.active_player,
        state.score_p0,
        state.score_p1,
        state.max_score,
        state.current_trick,
        state.trick_cards.copy(),
        state.trick_results.copy(),
        state.played_cards.copy(),
        state.trick_leader,
        state.envido_resolved,
        state.envido_chain.copy(),
        state.pending_envido_response,
        state.pending_envido_from,
        state.truco_level,
        state.truco_caller,
        state.pending_truco_response,
        state.pending_truco_from,
        state.is_hand_done,
        state.winner,
        state.points_won,
    )


def step(state: GameState, action: Action) -> GameState:
    new_state = _clone_state(state)
    player = new_state.active_player
    other_player = 1 - player

    if action == Action.IR_AL_MAZO:
        new_state.is_hand_done = True
        new_state.winner = other_player
        new_state.points_won = new_state.truco_level
        return new_state

    if action in [Action.ENVIDO, Action.REAL_ENVIDO, Action.FALTA_ENVIDO]:
        new_state.envido_chain.append(action)
        new_state.pending_envido_response = True
        new_state.pending_envido_from = other_player
        new_state.active_player = other_player
        return new_state

    if action == Action.QUIERO_ENVIDO:
        new_state.envido_chain.append(action)
        new_state.pending_envido_response = False
        new_state.envido_resolved = True
        new_state.active_player = (
            other_player if len(new_state.trick_cards) % 2 == 1 else new_state.trick_leader
        )

        envido_p0 = calculate_envido(new_state.hands[0])
        envido_p1 = calculate_envido(new_state.hands[1])
        winner = 0 if envido_p0 > envido_p1 else (1 if envido_p1 > envido_p0 else new_state.mano)

        if Action.FALTA_ENVIDO in new_state.envido_chain:
            pts = calculate_falta_envido_points(
                new_state.score_p0, new_state.score_p1, new_state.max_score
            )
        else:
            pts = 0
            for call in new_state.envido_chain:
                if call == Action.ENVIDO:
                    pts += 2
                elif call == Action.REAL_ENVIDO:
                    pts += 3
        if winner == 0:
            new_state.score_p0 = min(new_state.max_score, new_state.score_p0 + pts)
        else:
            new_state.score_p1 = min(new_state.max_score, new_state.score_p1 + pts)

        return new_state

    if action == Action.NO_QUIERO_ENVIDO:
        new_state.envido_chain.append(action)
        new_state.pending_envido_response = False
        new_state.envido_resolved = True
        new_state.active_player = (
            other_player if len(new_state.trick_cards) % 2 == 1 else new_state.trick_leader
        )

        caller = other_player
        chain = new_state.envido_chain
        if len(chain) == 2:
            pts = 1
        else:
            pts = 0
            for call in chain[:-2]:
                if call == Action.ENVIDO:
                    pts += 2
                elif call == Action.REAL_ENVIDO:
                    pts += 3
            if pts == 0:
                pts = 1

        if caller == 0:
            new_state.score_p0 += pts
        else:
            new_state.score_p1 += pts

        return new_state

    if action in [Action.TRUCO, Action.RETRUCO, Action.VALE_CUATRO]:
        new_state.pending_truco_response = True
        new_state.pending_truco_from = other_player
        new_state.truco_caller = player
        new_state.active_player = other_player
        if action == Action.VALE_CUATRO:
            new_state.truco_level = 3
        if not new_state.envido_resolved:
            new_state.envido_resolved = True
        return new_state

    if action == Action.QUIERO_TRUCO:
        new_state.pending_truco_response = False
        new_state.truco_level += 1
        new_state.active_player = (
            other_player if len(new_state.trick_cards) % 2 == 1 else new_state.trick_leader
        )
        return new_state

    if action == Action.NO_QUIERO_TRUCO:
        new_state.is_hand_done = True
        new_state.winner = other_player
        new_state.points_won = new_state.truco_level
        return new_state

    if 0 <= action.value <= 5:  # PLAY_CARD
        if not new_state.envido_resolved:
            new_state.envido_resolved = True

        card_idx = action.value % 3
        card = new_state.hands[player].pop(card_idx)
        new_state.trick_cards.append((player, card))
        new_state.played_cards.append((player, card))

        if len(new_state.trick_cards) == 2:
            p0_card = (
                new_state.trick_cards[0][1]
                if new_state.trick_cards[0][0] == 0
                else new_state.trick_cards[1][1]
            )
            p1_card = (
                new_state.trick_cards[1][1]
                if new_state.trick_cards[1][0] == 1
                else new_state.trick_cards[0][1]
            )

            res = compare_cards(p0_card, p1_card)
            new_state.trick_results.append(res)

            hr = resolve_hand(new_state.trick_results, new_state.mano)
            if hr is not None:
                new_state.is_hand_done = True
                new_state.winner = hr
                new_state.points_won = new_state.truco_level
            else:
                new_state.current_trick += 1
                if res != 0:
                    winner_player = 0 if res == 1 else 1
                    new_state.trick_leader = winner_player
                    new_state.active_player = winner_player
                else:
                    new_state.active_player = new_state.trick_leader
                new_state.trick_cards = []
        else:
            new_state.active_player = other_player

    return new_state


def infoset_key(state:GameState, player:int) -> tuple:
    """Flat, hashable tuple of everything a player can see."""
    envido_results:tuple[int | None, int] = (None, 0) # (envido winner, envido points)

    if state.envido_resolved:
        if state.score_p0 > 0:
            envido_results = (0, state.score_p0)
        elif state.score_p1 > 0:
            envido_results = (1, state.score_p1)
        else: envido_results = (None, 0)
    
    envido_winner = envido_results[0]
    envido_points = envido_results[1]

    infoset:tuple = (
        # the hand is sorted so same hands do not get repeated
        tuple(sorted(state.hands[player], key=lambda c: (c.number, c.suit))),
        tuple(state.played_cards),
        tuple(state.trick_results), tuple(state.trick_cards), state.trick_leader,
        state.mano,
        envido_winner, envido_points,
        tuple(state.envido_chain), state.envido_resolved,
        state.truco_level, state.truco_caller,
        state.pending_envido_from, state.pending_envido_response,
        state.pending_truco_from, state.pending_truco_response
    )

    return infoset

File: crates/truco_engine/src/state.rs
//! BitboardState: exactly 32 bytes zero-allocation copy state machine.

use crate::actions::*;
use crate::card::{compare_cards, CARD_RANKS};
use crate::isomorphism::canonicalize_hand;
use crate::rules::{calculate_envido, calculate_falta_envido_points, resolve_hand};

const FLAG_ENVIDO_RESOLVED: u8 = 1 << 0;
const FLAG_PENDING_ENVIDO: u8 = 1 << 1;
const FLAG_PENDING_TRUCO: u8 = 1 << 2;
const FLAG_IS_HAND_DONE: u8 = 1 << 3;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
#[repr(C)]
pub struct BitboardState {
    pub hands: [[u8; 3]; 2],     // 6 bytes (0..39 or 255 if played)
    pub trick_cards: [u8; 2],    // 2 bytes (cards in current trick, 255 if empty)
    pub trick_results: [i8; 3],  // 3 bytes (1 for P0, -1 for P1, 0 for parda, 127 if unplayed)
    pub trick_leader: u8,        // 1 byte
    pub current_trick: u8,       // 1 byte
    pub active_player: u8,       // 1 byte
    pub mano: u8,                // 1 byte
    pub score_p0: u8,            // 1 byte
    pub score_p1: u8,            // 1 byte
    pub max_score: u8,           // 1 byte
    pub envido_chain: [u8; 6],   // 6 bytes (actions in envido chain, 255 if empty)
    pub envido_chain_len: u8,    // 1 byte
    pub pending_envido_from: u8, // 1 byte (0, 1, or 255)
    pub truco_level: u8,         // 1 byte (1..4)
    pub truco_caller: u8,        // 1 byte (0, 1, or 255)
    pub pending_truco_from: u8,  // 1 byte (0, 1, or 255)
    pub flags: u8,               // 1 byte (packed booleans)
    pub winner: u8,              // 1 byte (0, 1, or 255)
    pub points_won: u8,          // 1 byte
}

// Compile-time assertion for exactly 32-byte size
const _: () = assert!(std::mem::size_of::<BitboardState>() == 32);

impl BitboardState {
    pub fn new(hands: [[u8; 3]; 2], mano: u8) -> Self {
        Self {
            hands,
            trick_cards: [255, 255],
            trick_results: [127, 127, 127],
            trick_leader: mano,
            current_trick: 0,
            active_player: mano,
            mano,
            score_p0: 0,
            score_p1: 0,
            max_score: 30,
            envido_chain: [255; 6],
            envido_chain_len: 0,
            pending_envido_from: 255,
            truco_level: 1,
            truco_caller: 255,
            pending_truco_from: 255,
            flags: 0,
            winner: 255,
            points_won: 0,
        }
    }

    #[inline(always)]
    pub fn active_player(&self) -> u8 {
        self.active_player
    }

    #[inline(always)]
    pub fn mano(&self) -> u8 {
        self.mano
    }

    #[inline(always)]
    pub fn trick_leader(&self) -> u8 {
        self.trick_leader
    }

    #[inline(always)]
    pub fn current_trick(&self) -> u8 {
        self.current_trick
    }

    #[inline(always)]
    pub fn truco_level(&self) -> u8 {
        self.truco_level
    }

    #[inline(always)]
    pub fn envido_resolved(&self) -> bool {
        (self.flags & FLAG_ENVIDO_RESOLVED) != 0
    }

    #[inline(always)]
    pub fn is_done(&self) -> bool {
        (self.flags & FLAG_IS_HAND_DONE) != 0
    }

    #[inline(always)]
    pub fn winner(&self) -> Option<u8> {
        if self.winner != 255 {
            Some(self.winner)
        } else {
            None
        }
    }

    #[inline(always)]
    pub fn points_won(&self) -> u8 {
        self.points_won
    }

    #[inline(always)]
    pub fn score(&self) -> (u8, u8) {
        (self.score_p0, self.score_p1)
    }

    pub fn legal_actions_mask(&self) -> u32 {
        if self.is_done() {
            return 0;
        }

        let player = self.active_player;

        if (self.flags & FLAG_PENDING_ENVIDO) != 0 {
            let last_call = self.envido_chain[(self.envido_chain_len - 1) as usize];
            if last_call == ACTION_ENVIDO {
                let envido_count = self.envido_chain[..self.envido_chain_len as usize]
                    .iter()
                    .filter(|&&a| a == ACTION_ENVIDO)
                    .count();
                if envido_count >= 2 {
                    return (1 << ACTION_QUIERO_ENVIDO)
                        | (1 << ACTION_NO_QUIERO_ENVIDO)
                        | (1 << ACTION_REAL_ENVIDO)
                        | (1 << ACTION_FALTA_ENVIDO);
                }
                return (1 << ACTION_QUIERO_ENVIDO)
                    | (1 << ACTION_NO_QUIERO_ENVIDO)
                    | (1 << ACTION_ENVIDO)
                    | (1 << ACTION_REAL_ENVIDO)
                    | (1 << ACTION_FALTA_ENVIDO);
            } else if last_call == ACTION_REAL_ENVIDO {
                return (1 << ACTION_QUIERO_ENVIDO)
                    | (1 << ACTION_NO_QUIERO_ENVIDO)
                    | (1 << ACTION_FALTA_ENVIDO);
            }
            return (1 << ACTION_QUIERO_ENVIDO) | (1 << ACTION_NO_QUIERO_ENVIDO);
        }

        if (self.flags & FLAG_PENDING_TRUCO) != 0 {
            let mut mask = (1 << ACTION_QUIERO_TRUCO)
                | (1 << ACTION_NO_QUIERO_TRUCO)
                | (1 << ACTION_IR_AL_MAZO);
            if self.truco_level == 1 {
                mask |= 1 << ACTION_RETRUCO;
            } else if self.truco_level == 2 {
                mask |= 1 << ACTION_VALE_CUATRO;
            }
            return mask;
        }

        let mut mask = 0u32;

        // Envido calls
        if self.current_trick == 0
            && !self.envido_resolved()
            && self.trick_cards[1] == 255
            && self.envido_chain_len == 0
        {
            mask |= (1 << ACTION_ENVIDO) | (1 << ACTION_REAL_ENVIDO) | (1 << ACTION_FALTA_ENVIDO);
        }

        // Truco raises
        if self.truco_level == 1 {
            mask |= 1 << ACTION_TRUCO;
        } else if self.truco_level == 2 && self.truco_caller != player {
            mask |= 1 << ACTION_RETRUCO;
        } else if self.truco_level == 3 && self.truco_caller != player {
            mask |= 1 << ACTION_VALE_CUATRO;
        }

        // Card plays
        let hand = &self.hands[player as usize];
        let remaining = hand.iter().filter(|&&c| c < 40).count();
        if remaining > 0 {
            mask |= (1 << ACTION_PLAY_CARD_0) | (1 << ACTION_PLAY_CARD_DOWN_0);
        }
        if remaining > 1 {
            mask |= (1 << ACTION_PLAY_CARD_1) | (1 << ACTION_PLAY_CARD_DOWN_1);
        }
        if remaining > 2 {
            mask |= (1 << ACTION_PLAY_CARD_2) | (1 << ACTION_PLAY_CARD_DOWN_2);
        }

        mask |= 1 << ACTION_IR_AL_MAZO;
        mask
    }

    pub fn step(&self, action: u8) -> Self {
        let mut next = *self;
        let player = next.active_player;
        let other = 1 - player;

        if action == ACTION_IR_AL_MAZO {
            next.flags |= FLAG_IS_HAND_DONE;
            next.winner = other;
            next.points_won = next.truco_level;
            return next;
        }

        if action == ACTION_ENVIDO || action == ACTION_REAL_ENVIDO || action == ACTION_FALTA_ENVIDO {
            if (next.envido_chain_len as usize) < next.envido_chain.len() {
                next.envido_chain[next.envido_chain_len as usize] = action;
                next.envido_chain_len += 1;
            }
            next.flags |= FLAG_PENDING_ENVIDO;
            next.pending_envido_from = other;
            next.active_player = other;
            return next;
        }

        if action == ACTION_QUIERO_ENVIDO {
            if (next.envido_chain_len as usize) < next.envido_chain.len() {
                next.envido_chain[next.envido_chain_len as usize] = action;
                next.envido_chain_len += 1;
            }
            next.flags &= !FLAG_PENDING_ENVIDO;
            next.flags |= FLAG_ENVIDO_RESOLVED;
            next.active_player = if next.trick_cards[0] != 255 && next.trick_cards[1] == 255 {
                other
            } else {
                next.trick_leader
            };

            let e0 = calculate_envido(&next.hands[0]);
            let e1 = calculate_envido(&next.hands[1]);
            let env_winner = if e0 > e1 {
                0
            } else if e1 > e0 {
                1
            } else {
                next.mano
            };

            let pts = if next.envido_chain[..next.envido_chain_len as usize].contains(&ACTION_FALTA_ENVIDO) {
                calculate_falta_envido_points(next.score_p0, next.score_p1, next.max_score)
            } else {
                let mut p = 0u8;
                for &a in &next.envido_chain[..next.envido_chain_len as usize] {
                    if a == ACTION_ENVIDO {
                        p += 2;
                    } else if a == ACTION_REAL_ENVIDO {
                        p += 3;
                    }
                }
                p
            };

            if env_winner == 0 {
                next.score_p0 = (next.score_p0 + pts).min(next.max_score);
            } else {
                next.score_p1 = (next.score_p1 + pts).min(next.max_score);
            }
            return next;
        }

        if action == ACTION_NO_QUIERO_ENVIDO {
            if (next.envido_chain_len as usize) < next.envido_chain.len() {
                next.envido_chain[next.envido_chain_len as usize] = action;
                next.envido_chain_len += 1;
            }
            next.flags &= !FLAG_PENDING_ENVIDO;
            next.flags |= FLAG_ENVIDO_RESOLVED;
            next.active_player = if next.trick_cards[0] != 255 && next.trick_cards[1] == 255 {
                other
            } else {
                next.trick_leader
            };

            let caller = other;
            let chain = &next.envido_chain[..next.envido_chain_len as usize];
            let pts = if chain.len() == 2 {
                1
            } else {
                let mut p = 0u8;
                for &a in &chain[..chain.len() - 2] {
                    if a == ACTION_ENVIDO {
                        p += 2;
                    } else if a == ACTION_REAL_ENVIDO {
                        p += 3;
                    }
                }
                if p == 0 {
                    1
                } else {
                    p
                }
            };

            if caller == 0 {
                next.score_p0 = (next.score_p0 + pts).min(next.max_score);
            } else {
                next.score_p1 = (next.score_p1 + pts).min(next.max_score);
            }
            return next;
        }

        if action == ACTION_TRUCO || action == ACTION_RETRUCO || action == ACTION_VALE_CUATRO {
            next.flags |= FLAG_PENDING_TRUCO;
            next.pending_truco_from = other;
            next.truco_caller = player;
            next.active_player = other;
            if action == ACTION_RETRUCO {
                next.truco_level = 2;
            } else if action == ACTION_VALE_CUATRO {
                next.truco_level = 3;
            }
            next.flags |= FLAG_ENVIDO_RESOLVED;
            return next;
        }

        if action == ACTION_QUIERO_TRUCO {
            next.flags &= !FLAG_PENDING_TRUCO;
            next.truco_level += 1;
            next.active_player = if next.trick_cards[0] != 255 && next.trick_cards[1] == 255 {
                other
            } else {
                next.trick_leader
            };
            return next;
        }

        if action == ACTION_NO_QUIERO_TRUCO {
            next.flags |= FLAG_IS_HAND_DONE;
            next.winner = other;
            next.points_won = next.truco_level;
            return next;
        }

        // Card plays (0..5)
        if action <= 5 {
            next.flags |= FLAG_ENVIDO_RESOLVED;
            let card_idx = (action % 3) as usize;
            let played_card = next.hands[player as usize][card_idx];

            // Shift remaining cards to keep valid cards contiguous at start of hands[player]
            let mut new_hand = [255u8; 3];
            let mut new_idx = 0;
            for i in 0..3 {
                if i != card_idx && next.hands[player as usize][i] < 40 {
                    new_hand[new_idx] = next.hands[player as usize][i];
                    new_idx += 1;
                }
            }
            next.hands[player as usize] = new_hand;

            if next.trick_cards[0] == 255 {
                next.trick_cards[0] = played_card;
                next.active_player = other;
            } else {
                next.trick_cards[1] = played_card;
                let leader = next.trick_leader;
                let p0_card = if leader == 0 {
                    next.trick_cards[0]
                } else {
                    played_card
                };
                let p1_card = if leader == 1 {
                    next.trick_cards[0]
                } else {
                    played_card
                };

                let res = compare_cards(p0_card, p1_card);
                let trick_idx = (next.current_trick as usize).min(2);
                next.trick_results[trick_idx] = res;

                let hr = resolve_hand(&next.trick_results[..=trick_idx], next.mano);
                if let Some(winner) = hr {
                    next.flags |= FLAG_IS_HAND_DONE;
                    next.winner = winner;
                    next.points_won = next.truco_level;
                } else if next.current_trick >= 2 {
                    next.flags |= FLAG_IS_HAND_DONE;
                    next.winner = next.mano;
                    next.points_won = next.truco_level;
                } else {
                    next.current_trick += 1;
                    if res != 0 {
                        let trick_winner = if res == 1 { 0 } else { 1 };
                        next.trick_leader = trick_winner;
                        next.active_player = trick_winner;
                    } else {
                        next.active_player = next.trick_leader;
                    }
                    next.trick_cards = [255, 255];
                }
            }
        }

        next
    }

    pub fn canonical_infoset_key(&self, player: u8) -> u64 {
        let hand_slice = &self.hands[player as usize];
        let canon_hand = canonicalize_hand(hand_slice);

        // ponytail: fast FNV-1a hash over canonical hand, trick rank history, and state variables
        let mut h: u64 = 0xcbf29ce484222325;
        macro_rules! feed {
            ($b:expr) => {
                h = (h ^ ($b as u64)).wrapping_mul(0x100000001b3);
            };
        }

        feed!(canon_hand[0]);
        feed!(canon_hand[1]);
        feed!(canon_hand[2]);

        // Trick cards on table normalized to rank
        for &tc in &self.trick_cards {
            if tc < 40 {
                feed!(CARD_RANKS[tc as usize]);
            } else {
                feed!(255u8);
            }
        }

        feed!(self.trick_results[0] as u8);
        feed!(self.trick_results[1] as u8);
        feed!(self.trick_results[2] as u8);

        feed!(self.trick_leader);
        feed!(self.current_trick);
        feed!(self.mano);

        feed!(self.flags);
        feed!(self.score_p0);
        feed!(self.score_p1);
        feed!(self.pending_envido_from);
        feed!(self.envido_chain_len);
        for i in 0..(self.envido_chain_len as usize).min(6) {
            feed!(self.envido_chain[i]);
        }

        feed!(self.truco_level);
        feed!(self.truco_caller);
        feed!(self.pending_truco_from);

        if h == 0 {
            1
        } else {
            h
        }
    }
}
